### PR TITLE
Add SAM 3 LiteText MobileCLIP lightweight text encoder

### PR DESCRIPTION
**Pull request by SimonZeng7108**
## Summary

This PR integrates **SAM 3 LiteText** — a drop-in lightweight text encoder for SAM 3 — based on [arXiv:2602.12173](https://arxiv.org/abs/2602.12173). It replaces the 353 M-parameter CLIP ViT-L text encoder with a compact MobileCLIP student encoder (42–124 M params depending on variant), while leaving the ViT-H image encoder and all other SAM 3 components unchanged.

### Motivation

The CLIP ViT-L text encoder is the memory bottleneck in SAM 3 for many real-world deployment scenarios (edge GPUs, multi-stream inference). LiteText cuts text-encoder params by **65–88 %** with comparable segmentation quality on SA-Co/Gold and LVIS benchmarks.

#### New files
- `ultralytics/models/sam/sam3/mobile_clip.py` — MobileCLIP text transformer backbone  
  - `MobileCLIPTextTransformer` with S0 (MCT/RepMixer), S1, and L variants  
  - `resize_pos_embed()` for context-length truncation after checkpoint loading  
  - `reparameterize()` on `MobileOneBlock` and `RepMixerBlock` for inference fusion  
- `ultralytics/models/sam/sam3/text_encoder_student.py` — SAM 3-compatible wrapper  
  - `TextStudentEncoder(cfg, context_length, output_dim)` — no extra tokenizer deps (uses `clip.tokenize()`)  
  - `set_context_length()` / forward signature matches `VETextEncoder` exactly  
  - Device auto-inferred from model parameters when not provided  

#### Modified files
- `ultralytics/models/sam/build_sam3.py`  
  - `_LITETEXT_CONFIGS` dict for S0 / S1 / L backbone configs  
  - `_detect_litetext_backbone(path)` — filename-based auto-detection (basename only, handles both naming conventions)  
  - `_detect_litetext_context_length(path)` — ctx16 / ctx32 from filename  
  - `build_sam3_image_model()` — auto-routes to `TextStudentEncoder` with no API breakage  
- `ultralytics/models/sam/sam3/__init__.py` — exports new classes  
- `docs/en/models/sam-3-litetext.md` — new documentation page  
- `mkdocs.yml` — nav entry added  

#### New tests
- `tests/test_sam3_litetext.py` — 36 tests across 5 classes  
  - Detection helpers, encoder unit tests, config coverage  
  - Architecture tests (monkeypatched `_load_checkpoint`, no weights required)  
  - Full predictor test (skipped when local checkpoint unavailable)  

### Checkpoint naming conventions supported

| Pattern | Example |
|---|---|
| Ultralytics | `sam3-litetext-s0.pt` |
| EfficientSAM3 original | `efficient_sam3_text_s0_ctx16_fixed.pt` |

### No breaking changes

`build_sam3_image_model()` and `SAM3SemanticPredictor` signatures are backwards-compatible. Standard `sam3.pt` continues to use `VETextEncoder` as before.

### Testing

```
pytest tests/test_sam3_litetext.py -v   # 36 passed, 0 failed
ruff check / ruff format                 # all files clean
```

End-to-end smoke test: SAM 3 LiteText S0 correctly segments a dog from `dog_person.jpeg` (0.956 confidence, 1 mask) at ~700 ms on RTX 4070 Ti.

---

Weights: [Simon7108528/EfficientSAM3](https://huggingface.co/Simon7108528/EfficientSAM3)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds SAM 3 LiteText support to Ultralytics with a lightweight MobileCLIP-based text encoder option, plus docs, navigation, and tests for auto-detected LiteText checkpoints. 🚀

### 📊 Key Changes
- Introduces `TextStudentEncoder` and `MobileCLIPTextTransformer` for SAM 3 LiteText integration as a lightweight replacement for the standard SAM 3 text encoder. 🧠
- Updates `build_sam3_image_model()` to auto-detect LiteText checkpoint variants from filenames and configure the correct backbone and context length automatically. ⚙️
- Adds LiteText backbone presets for S0, S1, and L variants, including context-length handling and post-load positional embedding resizing. 🪶
- Exposes new SAM 3 text encoder components through `ultralytics.models.sam.sam3.__init__`. 📦
- Adds a new model documentation page for SAM 3 LiteText and registers it in `mkdocs.yml` navigation. 📚
- Adds comprehensive tests covering filename detection, config correctness, encoder behavior, model wiring, and end-to-end predictor inference with local checkpoints. ✅

### 🎯 Purpose & Impact
- Enables lower-memory SAM 3 text-prompt segmentation workflows while preserving the existing `SAM3SemanticPredictor` interface. 💡
- Improves usability by letting users load LiteText checkpoints without manual backbone selection or extra configuration. 🙌
- Expands SAM 3 support for edge and memory-constrained deployments where the standard text encoder can be a bottleneck. 📱
- Strengthens maintainability and reliability through targeted unit and integration tests for the new LiteText pathway. 🛡️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SimonZeng7108, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Since this PR adds a new SAM 3 LiteText feature with docs and tests, please review the checklist below to help ensure smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/models/sam/build_sam3.py [hunk 1]
```diff
@@ -18,10 +20,89 @@
...
+    if "litetext" in name or "lite-text" in name:
+        if "litetext-l" in name or "litetext_l" in name:
+            return "L"
+        if "litetext-s1" in name or "litetext_s1" in name:
+            return "S1"
+        if "litetext-s0" in name or "litetext_s0" in name:
+            return "S0"
+        return None
+    # Original EfficientSAM3 convention: efficient_sam3_text_{s0,s1,l}_ctx*.pt
+    if "efficient_sam3_text" in name or "mobileclip" in name:
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/build_sam3.py]: ⚠️ **HIGH**: This is broad enough to classify files like `efficient_sam3_image_encoder_mobileclip_*` as full LiteText checkpoints. But `_load_checkpoint()` only loads `detector.*` keys, so if the file is actually an image-encoder-only checkpoint we silently return a mostly uninitialized SAM3 model because `strict=False`. Please narrow this detection to actual detector/text checkpoint names or validate the checkpoint contents before building the LiteText path.

File: ultralytics/models/sam/sam3/text_encoder_student.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    def __init__(self, cfg: dict, context_length: int, output_dim: int) -> None:
+        super().__init__()
+        self.context_length = context_length
+
+        # MobileCLIP text transformer (includes embedding layer and positional embedding).
+        # Override context_length in cfg so the positional embedding matches the checkpoint size.
+        # projection_dim=cfg["dim"] means the internal projection has the same dim; we apply
+        # our own projector below to map to SAM3's output_dim.
+        effective_cfg = {**cfg, "context_length": context_length}
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/sam3/text_encoder_student.py]: ⚠️ **HIGH**: This instantiates the MobileCLIP backbone with the *runtime* context length instead of the checkpoint's original context length (`cfg["context_length"]`, documented as 77). That means LiteText checkpoints with 77-position embeddings will hit a size mismatch during `_load_checkpoint()` before `set_context_length()` in `build_sam3_image_model()` can truncate them. The model needs to be built at the checkpoint length first, then resized after weights are loaded.

File: docs/en/models/sam-3-litetext.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+    No. Both are auto-detected from the checkpoint filename. You can override context length by passing `litetext_context_length` to `build_sam3_image_model()` directly.
+
+??? question "Can I use LiteText for video tracking?"
+
+    Yes. `SAM3VideoSemanticPredictor` inherits from `SAM3SemanticPredictor` and uses the same `build_sam3_image_model` path, so LiteText checkpoints work for video tracking without any code changes.
+
+??? question "Why does the S0 MCT architecture use RepMixer blocks?"
+
+    RepMixer blocks use a depthwise-separable structure that can be **re-parameterised** (fused) at inference time into a single depthwise convolution, reducing latency. Call `model.backbone.language_backbone.encoder.reparameterize()` after loading weights if you need maximum throughput.
```
Comment (UltralyticsAssistant) [inline docs/en/models/sam-3-litetext.md]: 💡 **MEDIUM**: `model.backbone.language_backbone.encoder` is a `MobileCLIPTextTransformer`, and this PR doesn't add a top-level `reparameterize()` method to that class. As written, this example will raise `AttributeError` for users. Either expose a public wrapper for reparameterization or remove this call from the docs.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice addition overall: the LiteText wiring, config presets, and test coverage are well structured. I found two implementation issues that can break real checkpoint loading/detection, plus one docs example that currently points users at a non-existent method.

💬 Posted 3 inline comments

**Comment by SimonZeng7108** [comment]
I have read the CLA Document and I sign the CLA

**Comment by SimonZeng7108** [comment]
@Laughing-q @Y-T-G @glenn-jocher @UltralyticsAssistant new PR opened for issue https://github.com/ultralytics/ultralytics/issues/24297.

**Comment by SimonZeng7108** [comment]
## Follow-up: All three code review issues addressed

All three `UltralyticsAssistant` inline comments have been resolved across commits [`b6f6e35`](https://github.com/SimonZeng7108/ultralytics/commit/b6f6e3519) and [`5257bc1`](https://github.com/SimonZeng7108/ultralytics/commit/5257bc193):

### 1. ⚠️ Detection too broad (`build_sam3.py`)
`efficient_sam3_image_encoder_mobileclip_*` filenames now return `None` from `_detect_litetext_backbone()` — the `"image_encoder"` substring guard ensures those image-encoder-only files are never routed through the LiteText path.

### 2. ⚠️ Context-length mismatch (`text_encoder_student.py`)
Fixed by adding `_peek_litetext_pos_embed_length()` which inspects the checkpoint before building the model. `TextStudentEncoder` is now constructed at the exact positional-embedding size stored in the checkpoint (whether 16, 32, or 77), eliminating the `load_state_dict` size mismatch in all cases. `set_context_length()` still handles truncation afterwards where needed.

### 3. 💡 Missing `reparameterize()` (`docs/en/models/sam-3-litetext.md`)
`reparameterize()` is now a proper public method on both `MobileCLIPTextTransformer` and `TextStudentEncoder`. The docs example updated to `model.backbone.language_backbone.reparameterize()` (one level shorter via the new wrapper).

---
All 36 tests pass locally (`pytest tests/test_sam3_litetext.py`).

**Comment by glenn-jocher** [comment]
Thanks for the update — the narrowed `_detect_litetext_backbone()` guard, checkpoint-aware context sizing via `_peek_litetext_pos_embed_length()`, and public `reparameterize()` wrapper all look like the right fixes; please keep the PR scoped to LiteText support only, and we’ll review the latest changes.

**Comment by glenn-jocher** [comment]
Thanks for the follow-up — those changes address the three review blockers, and keeping the LiteText path isolated and checkpoint-aware is the right direction, so please keep CI green and we’ll take the next review pass from here.

**Comment by SimonZeng7108** [comment]
> Thanks for the follow-up — those changes address the three review blockers, and keeping the LiteText path isolated and checkpoint-aware is the right direction, so please keep CI green and we’ll take the next review pass from here.

Thanks Glenn, let me know what is the best way to give you guys the weights. I haved train 6 new weights for different mobileclip variant and context window length.

**Comment by SimonZeng7108** [comment]
> > Thanks for the follow-up — those changes address the three review blockers, and keeping the LiteText path isolated and checkpoint-aware is the right direction, so please keep CI green and we’ll take the next review pass from here.
> 
> Thanks Glenn, let me know what is the best way to give you guys the weights. I haved train 6 new weights for different mobileclip variant and context window length.

Weights are uploaded to https://huggingface.co/Simon7108528/EfficientSAM3/tree/main/sam3_litetext - 6 weights starting with sam3_litetext_*

**Comment by glenn-jocher** [comment]
Thanks, hosting them externally in the [Hugging Face weights folder](https://huggingface.co/Simon7108528/EfficientSAM3/tree/main/sam3_litetext) is the right approach; please add the exact checkpoint names and a short loading note to the PR description or [SAM 3 docs](https://docs.ultralytics.com/models/sam-3/) so reviewers can validate the LiteText path without adding binaries to the repo.

**Comment by SimonZeng7108** [comment]
can anyone give some ideas how to proceed with this PR? @Y-T-G @Laughing-q Thank you!

**Comment by glenn-jocher** [comment]
Thanks @SimonZeng7108528 — there is genuine work here, and the distillation results you posted are interesting. I'm going to close it, and I want to give you the full reasoning rather than a one-liner.

**We already own MobileCLIP, and this adds a second copy of it.**

Ultralytics ships MobileCLIP today through `ultralytics/nn/text_model.py`:

- `MobileCLIP` (line 163) imports the `mobileclip` package and, if it is missing, installs **our own maintained fork**: `check_requirements("git+https://github.com/ultralytics/mobileclip.git")`.
- `MobileCLIPTS` (line 255) loads TorchScript text encoders — `mobileclip_blt.ts`, `mobileclip2_b.ts` — from **our own release assets** via `attempt_download_asset` (registered in `ultralytics/utils/downloads.py`).
- `build_text_model()` is the registry that owns this: it already dispatches `clip`, `mobileclip` and `mobileclip2`.

This PR instead vendors a 1,055-line copy of the MobileCLIP architecture into `ultralytics/models/sam/sam3/mobile_clip.py`. That is a parallel implementation of something the repo already has an owner for, and it would leave us maintaining two MobileCLIPs that drift apart. Our contributing rules put this first: extend the existing owner rather than creating a parallel abstraction, and search and reuse before adding.

**Scope.** The PR is +2093/-19 across 11 files, of which roughly 1,600 are non-documentation code lines. We treat anything over 50 net added non-documentation lines as a high-barrier exception that has to justify why it cannot be smaller or reuse an existing owner. A vendored third-party model file plus a 378-line test file does not clear that bar.

**Runtime dependency on assets we do not control.** The feature needs six 2.2-2.5 GB checkpoints from `huggingface.co/Simon7108528/EfficientSAM3`, a personal account. Everything we ship resolves weights from Ultralytics release assets, for exactly this reason — if that repo is renamed, emptied or rate-limited, the feature breaks for every user and we cannot fix it. Any weights we depend on have to be hosted on our side.

**Two specific things to carry forward if you revisit this:**

1. `_peek_litetext_pos_embed_length()` (`build_sam3.py:226`) calls `torch_load()` on the **entire** 2.2-2.5 GB checkpoint purely to read `shape[2]` of one positional-embedding tensor, then the checkpoint is loaded a second time for the real build. It is also wrapped in a bare `except Exception: pass`, so any real failure — a corrupt file, an OOM — is silently swallowed and the code proceeds with a wrong context length. The filename already encodes this: `_detect_litetext_context_length()` right below it reads `ctx16`/`ctx32` from the name without opening the file at all.
2. `vl_combiner.py` wraps `from torch.nn.attention import SDPBackend, sdpa_kernel` in a `try/except ImportError` with a no-op `sdpa_kernel` stub "for PyTorch < 2.0". That is an unrelated compatibility shim riding along in a feature PR, and the no-op version silently changes attention-backend selection rather than failing visibly.

**If you want to pursue the idea**, the shape that would get traction is much smaller: land the LiteText encoder in [`ultralytics/mobileclip`](https://github.com/ultralytics/mobileclip), register it as a variant in `build_text_model()`, and let us host the distilled weights as Ultralytics release assets. That reuses the owner, keeps the package diff to a handful of lines, and puts the checkpoints somewhere we can support. A distilled lightweight text encoder is a reasonable thing to want — it is the packaging here, not the idea, that does not fit.

Thanks again for the effort, and sorry for the slow turnaround on it.
